planner/core/exhaust_physical_plans.go

```diff
@@ -522,78 +576,337 @@ func (p *LogicalJoin) constructInnerIndexScan(ds *DataSource, idx *model.IndexIn
+	for i, opType := range cwc.OpType {
+		constantArg, err := cwc.opArg[i].Eval(row)
+		if err != nil {
+			return nil, err
+		}
+		cwc.tmpConstant[i].Value = constantArg
+		newExpr, err := expression.NewFunction(ctx, opType, types.NewFieldType(mysql.TypeTiny), cwc.targetCol, cwc.tmpConstant[i])
+		if err != nil {
+			return nil, err
 		}
+		exprs = append(exprs, newExpr)
+	}
+	ranges, err := ranger.BuildColumnRange(exprs, ctx.GetSessionVars().StmtCtx, cwc.targetCol.RetType, cwc.colLength)
+	if err != nil {
+		return nil, err
 	}
+	return ranges, nil
+}
 
-	return res.Ranges, append(remained, res.RemainedConds...), keyOff2IdxOff
+func (cwc *ColWithCompareOps) resolveIndices(schema *expression.Schema) (err error) {
+	for i := range cwc.opArg {
+		cwc.opArg[i], err = cwc.opArg[i].ResolveIndices(schema)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
-func (p *LogicalJoin) buildFakeEqCondsForIndexJoin(keys, idxCols []*expression.Column, colLengths []int,
-	innerFilters []expression.Expression) (accesses, eqConds, remained []expression.Expression, keyOff2IdxOff []int) {
-	// Check whether all join keys match one column from index.
-	keyOff2IdxOff = joinKeysMatchIndex(keys, idxCols, colLengths)
-	if keyOff2IdxOff == nil {
-		return nil, nil, nil, nil
+func (cwc *ColWithCompareOps) String() string {
+	buffer := bytes.NewBufferString("")
+	for i := range cwc.OpType {
+		buffer.WriteString(fmt.Sprintf("%v(%v, %v)", cwc.OpType[i], cwc.targetCol, cwc.opArg[i]))
+		if i < len(cwc.OpType)-1 {
+			buffer.WriteString(" ")
+		}
 	}
+	return buffer.String()
+}
 
-	usableKeys := make([]*expression.Column, 0, len(keys))
+func (ijHelper *indexJoinBuildHelper) checkIndex(innerKeys []*expression.Column, idxCols []*expression.Column, colLens []int) {
+	tmpSchema := expression.NewSchema(innerKeys...)
+	ijHelper.curIdxOff2KeyOff = make([]int, len(idxCols))
+	ijHelper.curPossibleUsedKeys = make([]*expression.Column, 0, len(idxCols))
+	ijHelper.curNotUsedIndexCols = make([]*expression.Column, 0, len(idxCols))
+	ijHelper.curNotUsedColLens = make([]int, 0, len(idxCols))
+	for i, idxCol := range idxCols {
+		ijHelper.curIdxOff2KeyOff[i] = tmpSchema.ColumnIndex(idxCol)
+		if ijHelper.curIdxOff2KeyOff[i] >= 0 {
+			ijHelper.curPossibleUsedKeys = append(ijHelper.curPossibleUsedKeys, idxCol)
+			continue
+		}
+		ijHelper.curNotUsedIndexCols = append(ijHelper.curNotUsedIndexCols, idxCol)
+		ijHelper.curNotUsedColLens = append(ijHelper.curNotUsedColLens, colLens[i])
+	}
+}
 
-	conds := make([]expression.Expression, 0, len(keys)+len(innerFilters))
-	eqConds = make([]expression.Expression, 0, len(keys))
-	// Construct a fake equal expression for every join key for calculating the range.
-	for i, key := range keys {
-		if keyOff2IdxOff[i] < 0 {
+func (ijHelper *indexJoinBuildHelper) analyzeSimpleFilters(innerPlan *DataSource, keyMatchedLen, matchedKeyCnt int) ([]expression.Expression, []expression.Expression, []expression.Expression) {
```

Comment:
how about changing this function name to `findAccessCondsOnUnusedIdxCols` and refactoring it to the following:
```go
// findAccessCondsOnUnusedIdxCols splits the filters on the inner DataSource
// into three parts:
// 1. accessOnUnusedIdxCols: access conditions(EQUAL/IN filters) on unused
//    index columns. accessOnUnusedIdxCols[i] stores the access condition on
//    the i-th unused index column.
// 2. filterOnIdxCols: filters on all the index columns which are covered by
//    access conditions.
// 3. otherFilters: other remaining filters.
func (ijHelper *indexJoinBuildHelper) findAccessCondsOnUnusedIdxCols(innerPlan *DataSource) (
    accessOnUnusedIdxCols, filterOnIdxCols, otherFilters []expression.Expression) {
    filterOnIdxCols = make([]expression.Expression, 0, len(innerPlan.pushedDownConds))
    filterOnOtherCols := make([]expression.Expression, 0, len(innerPlan.pushedDownConds))

    // Split all the filters to two parts:
    // 1. filters on index columns which are covered by Join ON conditions,
    //    stored in filterOnIdxCols.
    // 2. filters on other columns, stored in filterOnOtherCols.
    for _, innerFilter := range innerPlan.pushedDownConds {
        affectedCols := expression.ExtractColumns(innerFilter)
        if expression.ColumnSliceIsIntersect(affectedCols, ijHelper.curPossibleUsedKeys) {
            filterOnIdxCols = append(filterOnIdxCols, innerFilter)
            continue
        }
        filterOnOtherCols = append(filterOnOtherCols, innerFilter)
    }

    // Extract the access condition(EQUAL/IN filters) on unused index columns.
    // accessOnUnusedIdxCols[i] stores the access condition on the i-th unused
    // index column.
    accessOnUnusedIdxCols, filtersOnUnusedIdxCols, otherFilters, _ := ranger.ExtractEqAndInCondition(
        innerPlan.ctx, filterOnOtherCols, ijHelper.curNotUsedIndexCols, ijHelper.curNotUsedColLens)

    filterOnIdxCols = append(filterOnIdxCols, filtersOnUnusedIdxCols...)
    return accessOnUnusedIdxCols, filterOnIdxCols, otherFilters
}
```

Reply:
This method focus on eq/in conditions. So i think `findUsefulEqAndInConds` may be better?

Reply:
Okay.

Reply:
The third return value actually also contains filter on index columns since it can be used later.

Reply:
how about renaming the second return value to `filterOnUsedIdxCols`

Reply:
No, the second part not only contains the filters which contain the used index columns.
If the not used index columns have length, then the filter may need be left a copy to be evaled again to check it really matches.
So `filterOnUsedIdxCols` is just one part of the returned value.